In [20]:
import scipy.stats as stats
import numpy as np

# Unpooled two sample T-test

### Investigates whether the means of two independent data samples differ from one another. in a two sample t-test, the null hypothesis is that the means of both groups are the same.

In [21]:
threshold = '.05'
confidence_level = '95%'

In [22]:
#GPU training accuracy and loss at end of 100th epoch on NIPS, how was the data split and are the splits matching in file with the CPU splits?

acc_GPU = [89.28794113979299,
        87.65432098765432,
        88.19210624766181,
        89.0447686743983,
        88.17028307769048,
        88.68468636987156,
        89.39549819179449,
        88.39007357525877,
        88.47113106372365,
        88.88733009103379,
        88.54283576505799,
        88.52257139294176,
        88.19522384337199,
        89.15076692854471,
        88.56310013717422]

loss_GPU = [1.2841982543468475,
        1.4400151073932648,
        1.3928805887699127,
        1.2995456755161285,
        1.3862846791744232,
        1.3315181732177734,
        1.2454382181167603,
        1.3729434609413147,
        1.3651424795389175,
        1.3257267624139786,
        1.3586173355579376,
        1.3555727303028107,
        1.3874153345823288,
        1.279147669672966,
        1.3579276502132416]

# recorded times for training
# region
# 1 training time 74.426351 seconds
# 2 training time 63.349414 seconds
# 3 training time 62.190161 seconds
# 4 training time 115.643672 seconds
# 5 training time 109.095492 seconds
# 6 training time 111.923984 seconds
# 7 training time 64.468815 seconds
# 8 training time 54.612448 seconds
# 9 training time 57.284454 seconds
# 10 training time 53.336792 seconds
# 11 training time 55.611280 seconds
# 12 training time 59.712895 seconds
# 13 training time 54.212680 seconds
# 14 training time 57.649283 seconds
# 15 training time 61.556259 seconds
# endregion

times = [74.426351,
63.349414,
62.190161,
115.643672,
109.095492,
111.923984,
64.468815,
54.612448,
57.284454,
53.336792,
55.611280,
59.712895,
54.212680,
57.649283,
61.556259]

med_time_seconds = 61.556259
med_time_mins = '1 min 1 sec, 1.025938 min'

avg_time_seconds = 70.33826533333333
avg_time_mins = '1 min 10 sec, 1.172304 min'

len(acc_GPU),len(loss_GPU),len(times),np.median(times), np.mean(times)

(15, 15, 15, 61.556259, 70.33826533333333)

In [23]:
# CPU training accuracy and loss at end of 100th epoch on NIPS

acc_CPU = [ 87.48129606246948,
        88.34331035614014,
        89.07282948493958,
        89.06347155570984,
        89.82572555541992,
        89.96757864952087,
        86.55692934989929,
        89.21935558319092,
        89.6807610988617,
        89.22714591026306,
        89.62776064872742,
        89.83819484710693,
        89.00891542434692,
        88.54595422744751,
        89.88339900970459
        ]
        
# delete files in out and downsampled pickles else it outputs previous accuracy and loss.
loss_CPU = [1.4749480485916138,
        1.3730924874544144,
        1.2999057471752167,
        1.2973851561546326,
        1.2326992899179459,
        1.206596463918686,
        1.5558882355690002,
        1.332784652709961,
        1.2444825172424316,
        1.30466128885746,
        1.2385924607515335,
        1.2206190079450607,
        1.2771505862474442,
        1.3373953998088837,
        1.1882243752479553
        ]

# recorded times for training
# region
# 1 training time 669.396899 seconds
# 2 training time 870.183800 seconds
# 3 training time 660 mins 0 seconds
# 4 training time 665.073142 seconds
# 5 training time 679.237786 seconds
# 6 training time 783.128665 seconds
# 7 training time 807.589425 seconds 
# 8 training time 657.907862 seconds
# 9 training time 758.976197 seconds last 
# 10 training time 551.298181 seconds
# 11 training time 785.830169 seconds
# 12 training time 612.268949 seconds
# 13 training time 776.659075 seconds
# 14 training time 753.562039 seconds
# 15 training time 858.507515 seconds
# endregion 

times = [669.396899,
870.183800,
660,
665.073142,
679.237786,
783.128665,
807.589425,
657.907862,
758.976197,
551.298181,
785.830169,
612.268949,
776.659075,
753.562039,
858.507515]

avg_time_seconds = 725.9746469333332
avg_time_mins = '12 min 5 sec, 12.099577 min'

med_time_seconds = 753.562039
med_time_mins = '12 min 33 sec,12.559367 min'

len(acc_CPU),len(loss_CPU),len(times),np.median(times),np.mean(times)

(15, 15, 15, 753.562039, 725.9746469333332)

In [24]:
stats.ttest_ind(a = acc_GPU, b = acc_CPU, equal_var = False)

Ttest_indResult(statistic=-1.4881497017338858, pvalue=0.1519576341917487)

### The resulting test for both models training accuracy after 100 epochs yields a p-value of 0.1519, which means there is a 15% chance we would see sample data this far apart if the two groups tested are actually identical.
### Using a 95% confidence level we can fail to reject the null hypothesis, since the p-value is greater than the corresponding significance level of 5%. 

In [25]:
stats.ttest_ind(a = loss_GPU, b = loss_CPU, equal_var = False)

Ttest_indResult(statistic=1.3619096518411842, pvalue=0.18779477137344236)

### The resulting test for both models training loss after 100 epochs yields a p-value of 0.1877, which means there is a 19% chance we would see sample data this far apart if the two groups tested are actually identical.
### Using a 95% confidence level we can fail to reject the null hypothesis, since the p-value is greater than the corresponding significance level of 5%. 

### Average training time for GPU Model 1 min 10 sec <br> Average training time for CPU Model 12 min 33 sec <br> Median training time for GPU Model 1 min 1 sec <br> Median training time for CPU Model 12 min 5 sec <br>